src/main/java/com/haxwell/apps/quizki/services/UserServiceImpl.java

```diff
@@ -0,0 +1,125 @@
+package com.haxwell.apps.quizki.services;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import com.haxwell.apps.quizki.dtos.CreatedUserDTO;
+import com.haxwell.apps.quizki.dtos.UserCreateDTO;
+import com.haxwell.apps.quizki.entities.User;
+import com.haxwell.apps.quizki.entities.UserRole;
+import com.haxwell.apps.quizki.exceptions.UserFieldsNotUniqueException;
+import com.haxwell.apps.quizki.exceptions.UserRoleNotInDatabaseException;
+import com.haxwell.apps.quizki.exceptions.ValidationErrorData;
+import com.haxwell.apps.quizki.repositories.UserRepository;
+import com.haxwell.apps.quizki.repositories.UserRoleRepository;
+
+@Service
+public class UserServiceImpl implements UserService {
+	
+	@Autowired
+	private UserRepository uRepo;
+	@Autowired
+	private UserRoleRepository uroleRepo;
+	
+	
+	public UserServiceImpl(UserRepository uRepo, UserRoleRepository uroleRepo) {
+
+		this.uRepo = uRepo;
+		this.uroleRepo = uroleRepo;
+	}
+	
+	
+	public CreatedUserDTO  createNewUser(UserCreateDTO ucdto) {
+		
+		long nameCount = uRepo.countByName(ucdto.getName());
+		long emailCount = uRepo.countByEmail(ucdto.getEmail());
+		
+		if (emailCount != 0 || nameCount != 0) {
+			ValidationErrorData data = new ValidationErrorData();
+			if(emailCount != 0)
+				data.addFieldError("email", "email.not.unique");
+			if(nameCount != 0)
+				data.addFieldError("name", "name.not.unique");
+			
+			
+			throw new UserFieldsNotUniqueException(data);
+		}
+		
+		long role_id = ucdto.getRoleId();
+		Optional<UserRole> urole = uroleRepo.findById(role_id);
+		
+		if(!urole.isPresent()) {
+			ValidationErrorData data = new ValidationErrorData();
+			data.addFieldError("role", "role.not.in.database");
+			throw new UserRoleNotInDatabaseException(data);
+		}
+		
+		//Create a new User instance and map it from the ucdto DTO here
+		User user = new User();
+		
+		user.setName(ucdto.getName());
```

Comment:
I think a cleaner implementation would be to put these setter calls in the User object itself. So, basically you'd pass a `ucdto` object in the `User` constructor, and call the set() methods there.

Reply:
Yes, not requiring the `User_Role` in the client was the change that lead to this refactoring in the first place! Along with removing `Demographic` and `Enabled` which are set to defaults now. The problem you're describing about needing the `Role` is handled in the `Service` with our rule that it be passed by `id` and the defaults are handled there as well. In the future I see handling password hashing in it and maybe other functions like subscribers and Authorization. I like the concept of the service layer and how it moves the logic out of the controllers and out of the models and provides flexibility by extension for the future.